In [1]:
from load_data import load_raw_data
from clean_data import clean_data
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import re
import seaborn as sns
import plotly.express as px
import plotly.io as pio

In [2]:
pio.templates['plotly'].layout['autosize'] = False

#### Reading and cleaning the dataset

In [3]:
dataset=load_raw_data()

In [4]:
df=clean_data(dataset)

In [5]:
df['SeniorCitizen']=df['SeniorCitizen'].map({0:'No',1:'Yes'})

In [6]:
df.head()

,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,Female,No,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,Male,No,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.50,No
2,Male,No,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,Male,No,No,No,45,No,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,Female,No,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


#### Risk analysis: partner + dependents

In [7]:
df[(df["Partner"]=="Yes") & (df["Dependents"]=="Yes")]["Churn"].value_counts(normalize=True)*100

Churn
No     85.689655
Yes    14.310345
Name: proportion, dtype: float64

In [8]:
print(f"Percentage of total customers: {len(df[(df["Partner"]=="Yes") & (df["Dependents"]=="Yes")])/len(df)*100:.2f}%")

Percentage of total customers: 24.74%


In [9]:
df[(df["Partner"]=="Yes") & (df["Dependents"]=="Yes")].describe()

,tenure,MonthlyCharges,TotalCharges
count,1740.000000,1740.000000,1740.000000
mean,41.578161,61.070805,2746.914914
std,23.622272,31.649204,2358.492234
min,1.000000,18.700000,18.800000
25%,20.000000,24.687500,761.275000
50%,45.000000,65.125000,1833.275000
75%,65.000000,88.850000,4559.975000
max,72.000000,118.750000,8672.450000


In [10]:
df[(df["Partner"]=="No") & (df["Dependents"]=="No")]["Churn"].value_counts(normalize=True)*100

Churn
No     65.762195
Yes    34.237805
Name: proportion, dtype: float64

In [11]:
print(f"Percentage of total customers: {len(df[(df["Partner"]=="No") & (df["Dependents"]=="No")])/len(df)*100:.2f}%")

Percentage of total customers: 46.64%


In [12]:
df[(df["Partner"]=="No") & (df["Dependents"]=="No")].describe()

,tenure,MonthlyCharges,TotalCharges
count,3280.000000,3280.000000,3280.000000
mean,23.303354,62.983735,1610.702439
std,21.854290,28.925200,1899.021150
min,1.000000,18.250000,18.850000
25%,4.000000,36.600000,193.600000
50%,16.000000,69.700000,810.650000
75%,39.000000,86.000000,2404.325000
max,72.000000,118.650000,8547.150000


In [13]:
len(df[(df["Partner"]=="No") & (df["Dependents"]=="No")])

3280

#### Summary:
Possible high-risk group with 34.24% churn rate: combination No partner + No dependents formed of 3280 customers (46.6% of total customers).

- Customers with partners and dependents churn less than customers without both: 14.31% against 34.24%, and they represent 24.74% and 46.64% of the total customers, respectively. Customers with partner and dependents have higher tenure, higher total charges and similar monthly charges, when comparing to customers without partner nor dependents.

#### Risk analysis: fiber optics + tech support

In [14]:
df[df["InternetService"]=="Fiber optic"]["Churn"].value_counts(normalize=True)*100

Churn
No     58.107235
Yes    41.892765
Name: proportion, dtype: float64

In [15]:
print(f"Percentage of total customers: {len(df[df["InternetService"]=="Fiber optic"])/len(df)*100:.2f}%")

Percentage of total customers: 44.03%


In [16]:
df[(df["InternetService"]=="Fiber optic") & (df["TechSupport"]=="No")]["Churn"].value_counts(normalize=True)*100

Churn
No     50.627803
Yes    49.372197
Name: proportion, dtype: float64

In [17]:
len(df[(df["InternetService"]=="Fiber optic") & (df["TechSupport"]=="No")])

2230

In [18]:
print(f"Percentage of total customers: {len(df[(df["InternetService"]=="Fiber optic") & (df["TechSupport"]=="No")])/len(df)*100:.2f}%")

Percentage of total customers: 31.71%


In [19]:
df[(df["InternetService"]=="Fiber optic") & (df["TechSupport"]=="Yes")]["Churn"].value_counts(normalize=True)*100

Churn
No     77.367206
Yes    22.632794
Name: proportion, dtype: float64

In [20]:
print(f"Percentage of total customers: {len(df[(df["InternetService"]=="Fiber optic") & (df["TechSupport"]=="Yes")]["Churn"])/len(df)*100:.2f}%")

Percentage of total customers: 12.32%


#### Summary:
Possible high-risk group with 49.37% churn: combination Fiber Optic + No tech support formed of 2230 customers (31.71% of total customers).

- Customers with fiber optic and tech support churn less than customers with fiber optic and no tech support: 22.63% against 49.37%, and they represent 12.32% and 31.71% of the total customers, respectively.

#### Risk analysis: senior citizen + tech support

In [21]:
df[(df["SeniorCitizen"]=="Yes") & (df["InternetService"]!="No")]["Churn"].value_counts(normalize=True)*100

Churn
No     56.788991
Yes    43.211009
Name: proportion, dtype: float64

In [22]:
print(f"Percentage of total customers: {len(df[(df["SeniorCitizen"]=="Yes") & (df["InternetService"]!="No")])/len(df)*100:.2f}%")

Percentage of total customers: 15.50%


In [23]:
len(df[(df["SeniorCitizen"]=="Yes") & (df["InternetService"]!="No")])

1090

In [24]:
df[(df["SeniorCitizen"]=="Yes") & (df["InternetService"]!="No") & (df["TechSupport"]=="No")]["Churn"].value_counts(normalize=True)*100

Churn
Yes    50.60241
No     49.39759
Name: proportion, dtype: float64

In [25]:
len(df[(df["SeniorCitizen"]=="Yes") & (df["InternetService"]!="No") & (df["TechSupport"]=="No")])

830

In [26]:
print(f"Percentage of total customers: {len(df[(df["SeniorCitizen"]=="Yes")  & (df["InternetService"]!="No") & (df["TechSupport"]=="No")])/len(df)*100:.2f}%")

Percentage of total customers: 11.80%


In [27]:
df[(df["SeniorCitizen"]=="Yes") & (df["InternetService"]!="No") & (df["TechSupport"]=="Yes")]["Churn"].value_counts(normalize=True)*100

Churn
No     80.384615
Yes    19.615385
Name: proportion, dtype: float64

In [28]:
print(f"Percentage of total customers: {len(df[(df["SeniorCitizen"]=="Yes") & (df["InternetService"]!="No") & (df["TechSupport"]=="Yes")])/len(df)*100:.2f}%")

Percentage of total customers: 3.70%


#### Summary:
Possible high-risk group with 49.4% churn: combination Senior Citizen + No tech support formed of 1090 customers (15.5% of total customers).

- Regarding senior citizen customers with internet, the ones that have tech support churn less than the ones with no tech support: 19.6% against 49.4%, and they represent 3.7% and 11.8% of the total customers, respectively.

#### Extra analysis: Phone Service + Internet Service

Customers who subscribe to more than one service have a greater impact on the company's revenue and, as shown previously, exhibit a higher churn rate. Therefore, it is worthwhile to compare the characteristics of this subgroup with those of the overall customer population.

In [29]:
df_reduced=df[(df["PhoneService"]=="Yes") & (df["InternetService"]!="No")]
df_reduced.head()

,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
1,Male,No,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.50,No
2,Male,No,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
4,Female,No,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes
5,Female,No,No,No,8,Yes,Yes,Fiber optic,No,No,Yes,No,Yes,Yes,Month-to-month,Yes,Electronic check,99.65,820.50,Yes
6,Male,No,No,Yes,22,Yes,Yes,Fiber optic,No,Yes,No,No,Yes,No,Month-to-month,Yes,Credit card (automatic),89.10,1949.40,No


##### Comparing across each categorical feature

In [30]:
categorical_features=[col for col in df.columns if df[col].dtype=='object']

In [31]:
categorical_features

['gender',
 'SeniorCitizen',
 'Partner',
 'Dependents',
 'PhoneService',
 'MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'Contract',
 'PaperlessBilling',
 'PaymentMethod',
 'Churn']

In [32]:
#To enable direct comparisons of category proportions, we exclude the options listed below. For example, when comparing the proportions of customers using Fiber Optic and DSL internet services, 
#the overall customer population should include only customers with an internet service. Otherwise, the comparison would be distorted by customers without internet service. Under this approach, 
#for this particular example, the only difference between the two groups is that the overall customer population also includes customers who subscribe exclusively to internet service.

exclude = {
    "MultipleLines": "No phone service",
    "InternetService": "No",
    "OnlineSecurity": "No internet service",
    "OnlineBackup": "No internet service",
    "DeviceProtection": "No internet service",
    "TechSupport": "No internet service",
    "StreamingTV": "No internet service",
    "StreamingMovies": "No internet service",
}

for col in categorical_features:
    if col in exclude:
        value=exclude[col]
        df_trimmed=df[df[col]!=value]
        df_reduced_trimmed=df_reduced[df_reduced[col]!=value]
    else:
        df_trimmed=df
        df_reduced_trimmed=df_reduced

    comparison=pd.concat((df_trimmed[col].value_counts(normalize=True),df_reduced_trimmed[col].value_counts(normalize=True)),axis=1, keys=["All customers", "Subgroup"])*100
    display(comparison)

,All customers,Subgroup
gender,,
Male,50.469283,50.165563
Female,49.530717,49.834437


,All customers,Subgroup
SeniorCitizen,,
No,83.759954,79.594371
Yes,16.240046,20.405629


,All customers,Subgroup
Partner,,
No,51.749147,51.345199
Yes,48.250853,48.654801


,All customers,Subgroup
Dependents,,
No,70.150739,73.965232
Yes,29.849261,26.034768


,All customers,Subgroup
PhoneService,,
Yes,90.32992,100.0
No,9.67008,NaN


,All customers,Subgroup
MultipleLines,,
No,53.290302,45.633278
Yes,46.709698,54.366722


,All customers,Subgroup
InternetService,,
Fiber optic,56.16836,64.072848
DSL,43.83164,35.927152


,All customers,Subgroup
OnlineSecurity,,
No,63.443396,64.114238
Yes,36.556604,35.885762


,All customers,Subgroup
OnlineBackup,,
No,56.00508,55.732616
Yes,43.99492,44.267384


,All customers,Subgroup
DeviceProtection,,
No,56.132075,56.229305
Yes,43.867925,43.770695


,All customers,Subgroup
TechSupport,,
No,62.98984,63.721026
Yes,37.01016,36.278974


,All customers,Subgroup
StreamingTV,,
No,50.961538,49.917219
Yes,49.038462,50.082781


,All customers,Subgroup
StreamingMovies,,
No,50.453556,49.648179
Yes,49.546444,50.351821


,All customers,Subgroup
Contract,,
Month-to-month,55.105233,61.568709
Two year,23.961889,18.480960
One year,20.932878,19.950331


,All customers,Subgroup
PaperlessBilling,,
Yes,59.2719,69.06043
No,40.7281,30.93957


,All customers,Subgroup
PaymentMethod,,
Electronic check,33.631968,41.742550
Mailed check,22.810011,14.693709
Bank transfer (automatic),21.928328,22.102649
Credit card (automatic),21.629693,21.461093


,All customers,Subgroup
Churn,,
No,73.421502,67.177152
Yes,26.578498,32.822848


##### Tenure, Monthly Charges, Total Charges

In [33]:
df.describe()

,tenure,MonthlyCharges,TotalCharges
count,7032.000000,7032.000000,7032.000000
mean,32.421786,64.798208,2283.300441
std,24.545260,30.085974,2266.771362
min,1.000000,18.250000,18.800000
25%,9.000000,35.587500,401.450000
50%,29.000000,70.350000,1397.475000
75%,55.000000,89.862500,3794.737500
max,72.000000,118.750000,8684.800000


In [34]:
df_reduced.describe()

,tenure,MonthlyCharges,TotalCharges
count,4832.000000,4832.000000,4832.000000
mean,33.056705,81.761207,2902.470788
std,24.636200,18.306134,2415.564128
min,1.000000,42.900000,42.900000
25%,9.000000,69.787500,659.600000
50%,30.000000,82.500000,2350.625000
75%,56.000000,95.700000,4871.150000
max,72.000000,118.750000,8684.800000


#### Summary: 
Customers subscribed to more than one service form a subgroup with 32.8% churn rate: the combination Internet Service + Phone Service comprises 4832 customers (68.7% of total customers).

Compared with the overall customer population, this subgroup presents the following characteristics:

- No significant differences observed in the proportion of male/female, number of customers with partners, customers with/without online security, online backup, device protection, tech support, streaming TV and streaming Movies.

- Higher proportion of senior citizens in this group: 20.4% compared with 16.2% in the overall population

- Lower proportion of customers with dependents in this group: 26% compared with 29.8% in the overall population

- Higher proportion of customers with multiple lines: 54.4% compared with 46.7% in the overall population

- Higher proportion of fiber optic internet: 64%(Fiber)/36%(DSL) compared with 56%(Fiber)/44%(DSL) in the overall population

- There is a higher proportion of month to month contracts and lower 1y, 2y, (m-m/1y/2y): 61.6%/20%/18.5% compared with 55%/21%/24% in the overall population

- Higher proportion of paperless billing: 69% compared with 59% in the overall population

- The proportion of Electronic check payments grew, while mailed check payments reduced, they are 41.7% (Electronic check) 14.7% (Mailed check) compared with 33.6% (Electronic check) 22.8% (Mailed check) in the overall population

- Higher churn rate: 32.8% compared with 26.6% in the overall population

- Similar tenure average(median) 33(30) months compared with 32.4(29) months in the overall population

- Higher monthly charges average(median) \\$81.76(\\$82.5) compared with \\$64.8(\\$70.35) in the overall population

- Higher total charges average(median) \\$2902.5(\\$2350.6) compared with \\$2283.3(\\$1397.5) in the overall population

These observations motivate a more detailed analysis of the churn rate as a function of the number of services subscribed to by each customer. For a deeper investigation, see the Risk Analysis by Number of Services notebook.